File: Gayeong/src/week4/BaekJoon1240.java
package BFS;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class BaekJoon1240 {

    static int n;
    static int m;
    static ArrayList<Edge>[] graph;

    static class Edge {
        int v, weight;

        Edge(int v, int weight) {
            this.v = v;
            this.weight = weight;
        }
    }

    public static void main(String[] agrs) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        m = sc.nextInt();
        sc.nextLine();

        graph = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) graph[i] =  new ArrayList<>();


        for (int i = 0; i < n-1; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            int w = sc.nextInt();

            graph[u].add(new Edge(v, w));
            graph[v].add(new Edge(u, w));
        }

        for (int i = 0; i < m; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();

            System.out.println(bfs(u, v));
        }
    }

    static int  bfs(int start, int end) {
        Queue<Edge> queue = new LinkedList<>();
        queue.offer(new Edge(start, 0));
        int[] visited = new int[n + 1];
        visited[start] = 0;

        while (!queue.isEmpty()) {
            Edge cur = queue.poll();

            if (cur.v == end) return visited[cur.v];

            for (Edge edge : graph[cur.v]) {
                if (visited[edge.v] == 0) {
                    visited[edge.v] = edge.weight + visited[cur.v];
                    queue.offer(edge);
                }
            }
        }

        return -1;
    }


}

File: Heepyo/src/week4/BaekJoon1240.kt
package week4

import java.util.*

fun main() {
    val (n, m) = readln().split(" ").map(String::toInt)
    val graph = buildMap<Int, ArrayList<Pair<Int, Int>>> {
        repeat(n - 1) {
            val (a, b, weight) = readln().split(" ").map(String::toInt)

            computeIfAbsent(a) { _ -> ArrayList<Pair<Int, Int>>() }.add(b to weight)
            computeIfAbsent(b) { _ -> ArrayList<Pair<Int, Int>>() }.add(a to weight)
        }
    }

    for (i in 0 until m) {
        val (s, e) = readln().split(" ").map(String::toInt)

        println(bfs(start = s, end = e, graph))
    }
}

fun bfs(start: Int, end: Int, graph: Map<Int, ArrayList<Pair<Int, Int>>>): Int {
    val queue = ArrayDeque(listOf(start to 0))
    val visited = mutableSetOf<Int>()

    while (!queue.isEmpty()) {
        val (current, sumOfWeight) = queue.poll()

        if (current == end) return sumOfWeight
        visited.add(current)
        for ((next, weight) in graph[current] ?: emptyList()) {
            if (!visited.contains(next)) {
                queue.offer(next to sumOfWeight + weight)
            }
        }
    }
    return 0
}